utils_linux.go

```diff
@@ -217,19 +218,37 @@ func createPidFile(path string, process *libcontainer.Process) error {
 	return os.Rename(tmpName, path)
 }
 
-// XXX: Currently we autodetect rootless mode.
-func isRootless() bool {
-	return os.Geteuid() != 0
+func isRootless(context *cli.Context) (bool, error) {
+	if context != nil {
+		b, err := parseBoolOrAuto(context.GlobalString("rootless"))
+		if err != nil {
+			return false, err
+		}
+		if b != nil {
+			return *b, nil
+		}
+		// nil b stands for "auto detect"
+	}
+	// Even if os.Geteuid() == 0, it might still require rootless mode,
+	// especially when running within userns.
+	// So we use system.GetParentNSeuid() here.
+	//
+	// TODO(AkihiroSuda): how to support nested userns?
+	return system.GetParentNSeuid() != 0, nil
```

Comment:
would it be safer to assume "yes" everytime it is running in an userNS (i.e. doesn't have the full 0-4294967295 range available in /proc/self/uid_map)?

Reply:
This is actually a good point. I think doing both (and if either is true, we assume we are in rootless mode) would work as well. The main problem with the 4294967295 check is that you can create user namespaces that have a full mapping but still require rootless tricks (and as I said above there's no real way of checking if that is true).

Reply:
is that because it is still theoretically possible to setup the full range and have root mapped to some other user in the host namespace?  If I am not missing something, that sounds quite difficult to achieve as it requires an intermediate user namespace with at least two mappings (root in the intermediate namespace mapped to non root in the host) that still has enough privileges to setup the full range for the final userNS.

In any case, I fully agree that having both checks is safer

Reply:
> is that because it is still theoretically possible to setup the full range and have root mapped to some other user in the host namespace?

That is also possible, but it's actually much simpler than that. Even if you have a 1-to-1 mapping of all users in a new user namespace certain operations will still fail purely because you are not in `&init_user_ns` (such as `mknod(2)`) -- because capabilities are scoped to your user namespace and your host UID is irrelevant.

While this is an edge-case it means that in general we have no way of checking whether we are in a user namespace or the host user namespace. The closest thing we have is `ioctl(NS_GET_PARENT)` but returns `-EPERM` even if there is no parent -- I would argue that it should return `-ENOENT` in that case but @ebiederm might not agree.

Reply:
In the kernel the inode for the initial user namespace is hardcoded, and it seems like a reliable (and undocumented AFAICS) way to detect if we are running in the init userNS:

```
$ cat > userns_checker.c << EOF
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>

int main ()
{
  struct stat sb;
  if (stat ("/proc/self/ns/user", &sb) < 0)
    return -1;
  printf ("is main userns? %s\n", 0xEFFFFFFDU == sb.st_ino ? "yes" : "no");
  return 0;
}
EOF

$ gcc -o userns_checker userns_checker.c

$ ./userns_checker 
is main userns? yes
$ unshare -r ./userns_checker 
is main userns? no
```

What do you think?

Reply:
Rebased PR and added the 4294967295 check. 
The 0xEFFFFFFD check can be follow-up PR after discussion.

Reply:
As far as I can tell this is not guaranteed -- it just happens to be that `ns_alloc_inum` returns the same thing on each boot. But any kernel change or module that results in `proc_alloc_inum` being called before `&init_user_ns` is instantiated is going to result in that inode number changing.

Not to mention that (as Eric said) special inode numbers in general are not something you should depend on (a perfect example is "detecting" `chroot(2)` by the inode number of `/` -- something that always breaks every time someone tries it with a new filesystem because they all use different inode numbers for `/`).